server/src/main/java/org/elasticsearch/plugins/scanners/ExtensibleFileReader.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.plugins.scanners;
+
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.elasticsearch.xcontent.XContentType.JSON;
+
+public class ExtensibleFileReader {
+    private static final Logger logger = LogManager.getLogger(ExtensibleFileReader.class);
+
+    private String extensibleFile;
+
+    public ExtensibleFileReader(String extensibleFile) {
+        this.extensibleFile = extensibleFile;
+    }
+
+    public Map<String, String> readFromFile() {
+        Map<String, String> res = new HashMap<>();
+        // todo should it be BufferedInputStream ?
+        try (InputStream in = getClass().getResourceAsStream(extensibleFile)) {
```

Comment:
I was wondering if getClass().getClassLoader().getResourceAsStream could be used here (no preference, but wanted to see the difference..)
what I found is that: 
`getClass().getResourceAsStream("/org/elasticsearch/plugins/scanners/extensibles.json)` works without any modifications to module-info.java
but
`getClass().getClassLoader().getResourceAsStream("org/elasticsearch/plugins/scanners/extensibles.json")` 
only works when `    opens org.elasticsearch.plugins.scanners;` is added to module-info.java
this is mentioned in ClassLoader javadoc

> Additionally, and except for the special case where the resource has a name ending with ".class", this method will only find resources in packages of named modules when the package is opened unconditionally (even if the caller of this method is in the same module as the resource).

https://docs.oracle.com/javase/9/docs/api/java/lang/ClassLoader.html#getResource-java.lang.String-

Class's javadoc does not seem to be that restrictive

> Resources in named modules are subject to the rules for encapsulation specified in the Module getResourceAsStream method and so this method returns null when the resource is a non-".class" resource in a package that is not open to the caller's module.

@ChrisHegarty any ideas why the JDK did this that way? I suppose that Class#getResource is more preferred to Classloader#getResource ?

Reply:
@pgomulka The difference is clear from the javadoc/spec - j.l.Class::getResourceAsStream is caller-sensitive and determines the resource access based on the module of the immediate caller. Where as j.l.ClassLoader::getResourceAsStream is not caller-sensitive, therefore cannot determine resource access so behaves with no regard to the immediate caller. Ok, but why?

Generally speaking, caller-sensitive methods should not be extensible, as they pose problems for the implementing class, as well as opening the possibility to non-conformance. The method in j.l.Class is not extensible, so can be caller-sensitive. The method in ClassLoader is extensible, so is not caller-sensitive.